### ISSUE TITLE
We need z-index in stack

### ISSUE THREAD

**Issue by vytautas-pranskunas-** [flutter/flutter#70836]
Hello, yesterday I have reported spurious lack of functionality in flutter and without it I cannot complete game migration. I have tried all proposed ways proposed on stackoverflow, Flutter cummunity silent about solution also
I have found another similar feature request here. 
So I am asking to reopen issue:

https://github.com/flutter/flutter/issues/70758#issuecomment-729675908

And implement it adding zindex to positioned and animated Position widgets or creating new widget ZIndex

It will simplify layouting dranatically as well as games cretion

Thank you

**Comment by iapicca** [comment]
@vytautas-pranskunas- 
self-duplicating [issues](https://github.com/flutter/flutter/issues/70758), are not the best way to proceed,
next time please consider give more time to reopen a closed issue befor opening a new one

this seems to me a very limited use case for a feature request,
one could argue that you can already achieve what you describe using [Transform](https://api.flutter.dev/flutter/widgets/Transform-class.html) and [Matrix4](https://api.flutter.dev/flutter/vector_math_64/Matrix4-class.html) ([example](https://medium.com/flutter/perspective-on-flutter-6f832f4d912e))

I'm also a bit confused by the involvment of the [Stack](https://api.flutter.dev/flutter/widgets/Stack-class.html),
children are "piled" on Z axes already, I'm not sure [Positioned](https://api.flutter.dev/flutter/widgets/Positioned-class.html) is the best candidate for the role

imho your feature request should be a 3rd party package, rather than part of the main framework

**Comment by vytautas-pranskunas-** [comment]
I was not sure if closed tasks did get an attention. 
I dear to disagree a limitation of use because then we have to argue about limitation of use position: absolute and z-index usage on the web.

Stackoverflow answers are not working or involves anormous implementation complexity.

Do not understand why z-index might be irrelevant for framework. From my point of view it would dramatically simplify layouting.

**Comment by vytautas-pranskunas-** [comment]
I looked at example you gave - transforms are meant to scew and rotate objects. Would you be able to give me a code snippet that makes first container rendered above second container using matrix4D?

**Comment by vytautas-pranskunas-** [comment]
I see you put likes to your answers but hesitate to give a solution to existing problem. 
1. Rearanging list arrays involves big difficulties when it comes to working with 56 cards when each of them can be dragged and stacked into combinations... Also it has negavtive effects to animations
2. I could not made elevation working with Material using STack layout it does not react to elevation
3. I have just tried Matrix4 - i was not able to get item with index 0 rendered on top of item with index 1 in a list.

So i would be more then happy if you could show how to do this because i am bit stuck and i cannot get answer from anywhere.
Otherwise i keep thiking that this problem did not exist on windows phone, it can be really easily solved on the web using CSS, ReactNative and NativeScript.

I do elieve in Flutter as a 1# framework to develop corss platform app but in order it to happen likes is not enough - it would be more helpful to provide example (if it is so easy) rather then keep liking showing that framework can do this and dev who requested that is just stupid. Or maye you will face same chalange and recognize the need of this or that feature.

**Comment by iapicca** [comment]
@vytautas-pranskunas- 

> 1. Rearanging list arrays involves big difficulties when it comes to working with 56 cards when each of them can be dragged and stacked into combinations... Also it has negavtive effects to animations

reordering the children of a stack works very well in principle
here's a [minimal code sample](https://dartpad.dartlang.org/6cb9f6f76412c975d2f93d5de9396ec2)

I believe using key properly would minimize if not eliminate completely any inconvenience with animations

**Comment by vytautas-pranskunas-** [comment]
Reordering 9 arrays and total 56 cards which can be stacked in combinations meaning that you might want to stack together indexes 0,15,37,50 together into one place leaving other cards in their previous places and relative to each other.. But how you can calculate this if 5 indexes just had angee places?

I understand that you example where you switched 2 arrays it is easy... I also understand that theoretically it is possible. But guys - you do not want to understand that it brings anormous amount of complexity where should not be ANY problem. 

Z-index would solve all this emediatly.

**Comment by iapicca** [comment]
@vytautas-pranskunas- 

I wrote this for you, try it out, I've also updated the [code sample](https://dartpad.dartlang.org/6cb9f6f76412c975d2f93d5de9396ec2)
to see the code in action, 
it's a very minimal helper, but could make the task easier

<details>
<summary>elevation helper</summary>

```dart

extension Elevete on Widget {
  Elevated elevate(num i) => Elevation(
        child: this,
        elevation: i,
      );
}

abstract class Elevated {
  const Elevated({this.elevation});
  final num elevation;
  Widget call();
}

class Elevation extends Elevated {
  const Elevation({@required this.elevation, @required this.child});
  final num elevation;
  @protected
  final Widget child;

  @override
  Widget call() => child;
}

abstract class AxisZInterface<T extends Elevated> {
  const AxisZInterface(this.childrenZ);
  final List<T> childrenZ;

  List<Widget> get orderedChildren;
}

class StackZ extends Stack implements AxisZInterface<Elevation> {
  StackZ({@required this.childrenZ});
  final List<Elevation> childrenZ;

  @override
  List<Widget> get orderedChildren {
    childrenZ.sort((a, b) => a.elevation.compareTo(b.elevation));
    return <Widget>[for (Elevated e in childrenZ) e()];
  }

  @override
  List<Widget> get children {
    childrenZ.sort((a, b) => a.elevation.compareTo(b.elevation));
    return <Widget>[for (Elevated e in childrenZ) e()];
  }
}
```

</details>

**Comment by vytautas-pranskunas-** [comment]
Hello,

@iapicca  thanx for your effort. I have tried .sort option 2 days ago but animation was not working properly (because basically what you build for me is ...cards.orderBy((c) => x.zIndex).toList()) so looking at your example i understood that maybe something wrong was with animation - after moving one card all cards were rearanging. After reading your comment about keys i assigned key to card class animation became normal.
So having index is not so crucial in my case and your extension and StackZ could be part of framework reducing amount of questions given about z-index especially for those who comes from web.

Thanks for your help that helped me identify problem.

**Comment by derolf** [comment]
I have some Platform widgets in a stack and when trying to rendering them using Keys, they still get reinitialised. So zIndex would be great!

**Comment by adabru** [comment]
I want to paint a marking box behind a text but it shall still catch pointer-events before the text. Z-Index seems to be the easiest solution IMO. I know I could use OverlayEntry for that. Also Stack would work if I duplicate the marking widget into two widgets, one behind the text and the other in front of the text. For now I'm using blendMode.

**Comment by sooxt98** [comment]
I would be great if all of the widget could apply zIndex which gives the flexibly of rearranging UI

**Comment by mohamadlounnas** [comment]
try this [https://pub.dev/packages/indexed][1]


<img src="https://raw.githubusercontent.com/physia/kflutter/main/indexed/doc/assets/demo.gif">

Online [demo](https://dartpad.dev/?id=2bbadcf3554cf5d7222b62c639a18e5d&null_safety=true)
Video [demo](https://user-images.githubusercontent.com/22839194/131275765-9f1c037c-b4c5-4dfe-877f-14d5ee3f16e5.mp4)

```
    Indexer(
        children: [
            Indexed(IndexedInterface
              // prefer add key when use AnimatedPositioned
              key: UniqueKey(),
              index: 100,
              child: Positioned(
                //...
              )
            ),
        ],
    );
```

**Comment by duck-dev-go** [comment]
+1

**Comment by yingming25** [comment]
Why not wrap your widget with [Material](https://api.flutter.dev/flutter/material/Material-class.html) then change elevation value?

**Comment by MuhmdHsn313** [comment]
any updates?

**Comment by lukas-pierce** [comment]
> any updates?

Yes, this package solves the problem [elevated_flex](https://pub.dev/packages/elevated_flex)

**Comment by vytautas-pranskunas-** [comment]
@lukas-pierce not exactly, you can manipulate index with conditions easily and have multiple elevated ekements above each other. With this package, as i understand, you can have only one above others. But it is better than nothing

**Comment by lukas-pierce** [comment]
> @lukas-pierce not exactly, you can manipulate index with conditions easily and have multiple elevated ekements above each other. With this package, as i understand, you can have only one above others. But it is better than nothing

Yes, this package only for elevate relative items in column or row.

But as far as I understand from the statement of your problem, you need a Flow widget; which allows you to describe at a lower level the position and order of widgets in layers

https://youtu.be/NG6pvXpnIso?si=ylhxOxhETLZqUBZV

**Comment by vytautas-pranskunas-** [comment]
Yes thats may be good solution, but i have long time ago solved that
problem in my game somehow.

**Comment by shishir4** [comment]
@vytautas-pranskunas- care to share?
